::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: Copyright (c) 2017 Enea AB and others.
:
: All rights reserved. This program and the accompanying materials
: are made available under the terms of the Apache License, Version 2.0
: which accompanies this distribution, and is available at
: http://www.apache.org/licenses/LICENSE-2.0
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Sun, 20 Aug 2017 22:41:26 +0200
Subject: [PATCH] mcp: salt-formulas: armband: Extend libvirt_domain

Add new state in armband salt formula that extends salt's virt
libvirt_domain.jinja template with support for:
- hw_firmware_type;
- virt_machine_model;
- cpu_model;

These will later be leveraged via salt virt formula with AArch64
specific values.

Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 mcp/config/states/virtual_control_plane            |  1 +
 .../armband/files/libvirt_domain.jinja.diff        | 46 ++++++++++++++++++++++
 .../armband/libvirt_domain_template.sls            |  4 ++
 3 files changed, 51 insertions(+)
 create mode 100644 mcp/salt-formulas/armband/files/libvirt_domain.jinja.diff
 create mode 100644 mcp/salt-formulas/armband/libvirt_domain_template.sls

diff --git a/mcp/config/states/virtual_control_plane b/mcp/config/states/virtual_control_plane
index 044e645..fb1c439 100755
--- a/mcp/config/states/virtual_control_plane
+++ b/mcp/config/states/virtual_control_plane
@@ -27,6 +27,7 @@ wait_for 90 "! salt -C 'kvm* or cmp*' test.ping | " \
 salt -C '* and not cfg01* and not mas01*' state.apply linux,ntp

 salt -C 'kvm*' state.sls armband || true
+salt -C 'kvm*' state.sls armband.libvirt_domain_template || true
 wait_for 5 "salt -C 'kvm*' state.sls libvirt"

 salt -C '* and not cfg01* and not mas01*' state.apply salt
diff --git a/mcp/salt-formulas/armband/files/libvirt_domain.jinja.diff b/mcp/salt-formulas/armband/files/libvirt_domain.jinja.diff
new file mode 100644
index 0000000..4e7fad5
--- /dev/null
+++ b/mcp/salt-formulas/armband/files/libvirt_domain.jinja.diff
@@ -0,0 +1,46 @@
+From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
+Date: Sun Aug 20 18:18:53 2017 +0200
+Subject: [PATCH] libvirt_domain.jinja: Add AArch64 support
+
+Salt virt state relies on a Jinja template to create a libvirt
+XML definition for each new VM.
+This template needs to be extended with a few specific options
+for AArch64:
+- UEFI loader support (pflash);
+- custom machine model (e.g. 'virt-2.9'), since AArch64 defaults to
+  'integratorcp';
+- custom cpu model;
+
+Allow all these to be parametrized from the salt virt formula,
+which we will also adapt to allow the params to be passed via our
+reclass model.
+
+Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
+---
+
+diff --git a/usr/lib/python2.7/dist-packages/salt/templates/virt/libvirt_domain.jinja b/usr/lib/python2.7/dist-packages/salt/templates/virt/libvirt_domain.jinja
+--- a/usr/lib/python2.7/dist-packages/salt/templates/virt/libvirt_domain.jinja
++++ b/usr/lib/python2.7/dist-packages/salt/templates/virt/libvirt_domain.jinja
+@@ -3,11 +3,22 @@
+         <vcpu>{{ cpu }}</vcpu>
+         <memory unit='KiB'>{{ mem }}</memory>
+         <os>
++                {% if custom_virt_machine %}
++                <type machine='{{ virt_machine_model }}'>hvm</type>
++                {% else %}
+                 <type>hvm</type>
++                {% endif %}
++                {% if os_loader_type == 'pflash' %}
++                <loader readonly='yes' type='{{ os_loader_type }}'>{{ os_loader }}</loader>
++                <nvram>{{ os_loader_nvram }}</nvram>
++                {% endif %}
+                 {% for dev in boot_dev %}
+                 <boot dev='{{ dev }}' />
+                 {% endfor %}
+         </os>
++        {% if cpu_mode == 'custom' %}
++        <cpu mode='{{ cpu_model }}' check='none'/>
++        {% endif %}
+         <devices>
+                 {% for diskname, disk in disks.items() %}
+                 <disk type='file' device='disk'>
diff --git a/mcp/salt-formulas/armband/libvirt_domain_template.sls b/mcp/salt-formulas/armband/libvirt_domain_template.sls
new file mode 100644
index 0000000..7cc8e22
--- /dev/null
+++ b/mcp/salt-formulas/armband/libvirt_domain_template.sls
@@ -0,0 +1,4 @@
+/usr/lib/python2.7/dist-packages/salt/templates/virt/libvirt_domain.jinja:
+  file.patch:
+  - source:  salt://armband/files/libvirt_domain.jinja.diff
+  - hash: "9d02cd8aa04497c2d75c72840a5c6c41"
